From 2e6463ea850e31a8c3992d1d9e1a1e86f12d0643 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Guido=20G=C3=BCnther?= <agx@sigxcpu.org>
Date: Mon, 19 Oct 2020 15:16:30 +0200
Subject: [PATCH 1/3] shell: Create toplevel manager early
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

So far we were lucky that no wl_display_roundtrip () would let
us miss existing toplevels.

Signed-off-by: Guido Günther <guido.gunther@puri.sm>
---
 src/shell.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/shell.c b/src/shell.c
index fa728d01..418e2d71 100644
--- a/src/shell.c
+++ b/src/shell.c
@@ -569,6 +569,10 @@ phosh_shell_constructed (GObject *object)
 
   G_OBJECT_CLASS (phosh_shell_parent_class)->constructed (object);
 
+  /* We bind this early since a wl_display_roundtrip () would make us miss
+     exising toplevels */
+  priv->toplevel_manager = phosh_toplevel_manager_new ();
+
   priv->transform = -1; /* force initial update */
   priv->monitor_manager = phosh_monitor_manager_new ();
   g_signal_connect_swapped (priv->monitor_manager,
@@ -601,7 +605,6 @@ phosh_shell_constructed (GObject *object)
   priv->lockscreen_manager = phosh_lockscreen_manager_new ();
   priv->idle_manager = phosh_idle_manager_get_default();
 
-  priv->toplevel_manager = phosh_toplevel_manager_new ();
   priv->faders = g_ptr_array_new_with_free_func ((GDestroyNotify) (gtk_widget_destroy));
 
   phosh_system_prompter_register ();
-- 
2.26.2


From 36cb33494dc36fa14fe82ed91b2c9faeb3c0c77f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Guido=20G=C3=BCnther?= <agx@sigxcpu.org>
Date: Mon, 19 Oct 2020 12:41:55 +0200
Subject: [PATCH 2/3] shell: Process all pending wayland events on startup
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This is needed to get reliable monitor information.

Signed-off-by: Guido Günther <guido.gunther@puri.sm>
---
 src/shell.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/shell.c b/src/shell.c
index 418e2d71..6c069881 100644
--- a/src/shell.c
+++ b/src/shell.c
@@ -580,6 +580,9 @@ phosh_shell_constructed (GObject *object)
                             G_CALLBACK (on_monitor_removed),
                             self);
 
+  /* Make sure we processed wayland protocol events */
+  phosh_wayland_roundtrip (phosh_wayland_get_default ());
+
   if (phosh_monitor_manager_get_num_monitors(priv->monitor_manager)) {
     PhoshMonitor *monitor = phosh_monitor_manager_get_monitor (priv->monitor_manager, 0);
     /* Can't invoke phosh_shell_set_primary_monitor () since the shell
-- 
2.26.2


From fb2769a359c6454bc09e7b63e6f28ac2fd639c71 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Guido=20G=C3=BCnther?= <agx@sigxcpu.org>
Date: Mon, 19 Oct 2020 12:42:39 +0200
Subject: [PATCH 3/3] shell: Pick correct built in monitor
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Simplify the code now that we know we have all the monitor
information

FIXME: this breaks getting the initial list of toplevels

Closes: #392

Signed-off-by: Guido Günther <guido.gunther@puri.sm>
---
 src/shell.c | 14 ++++++--------
 1 file changed, 6 insertions(+), 8 deletions(-)

diff --git a/src/shell.c b/src/shell.c
index 6c069881..501fd1e3 100644
--- a/src/shell.c
+++ b/src/shell.c
@@ -580,26 +580,24 @@ phosh_shell_constructed (GObject *object)
                             G_CALLBACK (on_monitor_removed),
                             self);
 
-  /* Make sure we processed wayland protocol events */
+  /* Make sure all outputs are up to date */
   phosh_wayland_roundtrip (phosh_wayland_get_default ());
 
   if (phosh_monitor_manager_get_num_monitors(priv->monitor_manager)) {
-    PhoshMonitor *monitor = phosh_monitor_manager_get_monitor (priv->monitor_manager, 0);
+    priv->builtin_monitor = phosh_shell_get_builtin_monitor (self);
+
+    g_debug ("Builtin monitor is %s, %d", priv->builtin_monitor->name,
+             phosh_monitor_is_configured (priv->builtin_monitor));
     /* Can't invoke phosh_shell_set_primary_monitor () since the shell
        object does not really exist yet but we need the primary monitor
        early for the panels */
-    priv->primary_monitor = g_object_ref (monitor);
+    priv->primary_monitor = g_object_ref (priv->builtin_monitor);
     g_signal_connect_swapped (priv->primary_monitor,
                               "configured",
                               G_CALLBACK (on_primary_monitor_configured),
                               self);
   }
 
-  if (phosh_monitor_is_builtin(priv->primary_monitor))
-    priv->builtin_monitor = priv->primary_monitor;
-  else
-    priv->builtin_monitor = phosh_shell_get_builtin_monitor(self);
-
   gtk_icon_theme_add_resource_path (gtk_icon_theme_get_default (),
                                     "/sm/puri/phosh/icons");
   css_setup (self);
-- 
2.26.2

